File: Practice/srm_500_hard_ProductAndSum.cpp
// https://community.topcoder.com/stat?c=problem_statement&pm=11345
#include <vector>
#include <stdio.h>
using ll = long long;
using namespace std;
ll pow(int x,int p,int mod)
{
  if (p==0)
    return 1;
  ll t = pow(x,p/2,mod);
  t=(t*t)%mod;
  if (p%2==1) t=(t*x)%mod;
  return t;
}
ll Fact[2501], InvFact[2501], Ones[2501];
ll M = 500500573;
struct ProductAndSum
{
  int getSum(int p2, int p3, int p5, int p7, int S)
  {
    Fact[0] = 1, Ones[0] = 0, InvFact[0] = 1;
    for (int i=1;i<=2500;i++)
    {
      Fact[i] = (Fact[i-1] * i) % M;
      Ones[i] = (Ones[i-1] * 10 + 1) % M;
      InvFact[i] = pow(Fact[i], M-2, M);
    }
    int result = 0;
    for (int u4=0;u4*2<=p2;u4++) for (int u8=0;u4*2+u8*3<=p2;u8++) for (int u9=0;u9*2<=p3;u9++)
      for (int u6=0;u4*2+u8*3+u6<=p2 && u9*2+u6<=p3;u6++)
      {
        int u2 = p2-u4*2-u8*3-u6;
        int u3 = p3-u9*2-u6;
        vector<int> u = {0,0,u2,u3,u4,p5,u6,p7,u8,u9};
        int s = 0, digits = 0; 
        for (int i=2;i<=9;i++) 
        {
          s += i*u[i];
          digits += u[i];
        }
        u[1] = S-s;
        if (u[1] < 0) continue;
        digits += u[1];
        ll tmp = Fact[digits-1];
        for (int i=1;i<=9;i++) tmp = (tmp * InvFact[u[i]]) % M;
        tmp = (tmp * S) % M;
        tmp = (tmp * Ones[digits]) % M;
        result = (result + tmp) %M;
      }
    return result;
  }
};

int main()
{
  //170475026
  printf("%d\n",ProductAndSum().getSum(2,0,0,0,4));
  printf("%d\n",ProductAndSum().getSum(0,0,0,0,10));
  printf("%d\n",ProductAndSum().getSum(2,0,0,0,5));
  printf("%d\n",ProductAndSum().getSum(1,1,1,1,10));
  printf("%d\n",ProductAndSum().getSum(5,5,5,5,100));
  printf("%d\n",ProductAndSum().getSum(100,100,100,100,2500));
}

File: Practice/srm_500_hard_ProductAndSum.py
# https://community.topcoder.com/stat?c=problem_statement&pm=11345
# in real, it would be failed (TLE, 3.6s for the worst case)
def measure(func):
  import datetime
  def measure_call(*args, **kwargs):
    begin = datetime.datetime.now()
    func(*args, **kwargs)
    print("Elapsed = %f secs" % ((datetime.datetime.now() - begin).total_seconds()))
  return measure_call

@measure
def solve(l2,l3,l5,l7,S):
  M = 500500573
  f = [1]
  ones = [0]
  for i in range(1,2501):
    f.append( (f[-1] * i) % M )
    ones.append((ones[-1] * 10 + 1) % M)
  def pow(x,p,mod):
    if p==0:
      return 1
    t = pow(x,p//2,mod) ** 2
    if p%2==1:
      t*=x
    return t%mod
  inv_f = [ pow(x, M-2, M) for x in f ]
  z = 0
  c = [0]*10
  l = [100]*10
  l[2],l[3],l[5],l[7] = l2,l3,l5,l7
  c[5],c[7] = l5,l7
  subs = [0] * 5
  subf = [0] * 5
  subs[0] = c[5] * 5 + c[7] * 7
  subf[0] = (inv_f[c[5]] * inv_f[c[7]]) % M
  def update(idx, val):
    subs[idx] = subs[idx-1] + c[val] * val
    subf[idx] = (subf[idx-1] * inv_f[c[val]]) % M
    return subs[idx] > S
  for c[6] in range(min(l[2],l[3])+1):
    if update(1,6): break
    for c[4] in range((l[2]-c[6])//2+1):
      if update(2,4): break
      for c[8] in range((l[2]-c[4]*2-c[6])//3+1):
        if update(3,8): break
        for c[9] in range((l[3]-c[6])//2+1):
          c[2] = l[2] - (c[4]*2+c[8]*3+c[6])
          c[3] = l[3] - (c[9]*2+c[6])
          if c[2] < 0 or c[3] < 0:
            break

          subs[4] = subs[3] + c[9] * 9 + c[2] * 2 + c[3] * 3
          c[1] = S - subs[4]
          if c[1] < 0:
            break

          subf[4] = (subf[3] * inv_f[c[9]]) % M
          subf[4] *= (inv_f[c[2]] * inv_f[c[3]] * inv_f[c[1]]) % M
          subf[4] %= M

          digits = sum(c)
          xsum = S * subf[4] * f[digits-1] * ones[digits]
          xsum %= M
          z += xsum
          z %= M
  print (z%M)
'''
solve(2,0,0,0,4)
solve(0,0,0,0,10)
solve(2,0,0,0,5)
solve(1,1,1,1,10)
solve(5,5,5,5,100)
'''
solve(100,100,100,100,2500) #170475026
